crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,723 @@
+            operation_entry.operation_id, operation_entry.child_id
+        );
+        Ok(())
+    }
+
+    async fn publish_c8y_executing_message(
+        &mut self,
+        child_id: &str,
+        message_box: &mut FirmwareManagerMessageBox,
+    ) -> Result<(), FirmwareManagementError> {
+        let c8y_child_topic = Topic::new_unchecked(
+            &C8yTopic::ChildSmartRestResponse(child_id.to_string()).to_string(),
+        );
+        let executing_msg = MqttMessage::new(
+            &c8y_child_topic,
+            DownloadFirmwareStatusMessage::status_executing()?,
+        );
+        message_box.send(executing_msg.into()).await?;
+        Ok(())
+    }
+
+    async fn publish_c8y_successful_message(
+        &mut self,
+        child_id: &str,
+        message_box: &mut FirmwareManagerMessageBox,
+    ) -> Result<(), FirmwareManagementError> {
+        let c8y_child_topic = Topic::new_unchecked(
+            &C8yTopic::ChildSmartRestResponse(child_id.to_string()).to_string(),
+        );
+        let successful_msg = MqttMessage::new(
+            &c8y_child_topic,
+            DownloadFirmwareStatusMessage::status_successful(None)?,
+        );
+        message_box.send(successful_msg.into()).await?;
+        Ok(())
+    }
+
+    async fn publish_c8y_failed_message(
+        &mut self,
+        child_id: &str,
+        failure_reason: &str,
+        message_box: &mut FirmwareManagerMessageBox,
+    ) -> Result<(), FirmwareManagementError> {
+        let c8y_child_topic = Topic::new_unchecked(
+            &C8yTopic::ChildSmartRestResponse(child_id.to_string()).to_string(),
+        );
+        let failed_msg = MqttMessage::new(
+            &c8y_child_topic,
+            DownloadFirmwareStatusMessage::status_failed(failure_reason.to_string())?,
+        );
+        message_box.send(failed_msg.into()).await?;
+        Ok(())
+    }
+
+    async fn publish_c8y_installed_firmware_message(
+        &mut self,
+        operation_entry: &FirmwareOperationEntry,
+        message_box: &mut FirmwareManagerMessageBox,
+    ) -> Result<(), FirmwareManagementError> {
+        let c8y_child_topic = Topic::new_unchecked(
+            &C8yTopic::ChildSmartRestResponse(operation_entry.child_id.clone()).to_string(),
+        );
+        let installed_firmware_payload = format!(
+            "115,{},{},{}",
+            operation_entry.name, operation_entry.version, operation_entry.server_url
+        );
+        let installed_firmware_message =
+            MqttMessage::new(&c8y_child_topic, installed_firmware_payload);
+        message_box.send(installed_firmware_message.into()).await?;
+        Ok(())
+    }
+
+    fn remove_entry_from_active_operations(
+        &mut self,
+        operation_key: &OperationKey,
+    ) -> ActiveOperationState {
+        if let Some(operation_state) = self.active_child_ops.remove(operation_key) {
+            operation_state
+        } else {
+            ActiveOperationState::Pending
```

Comment:
Falling back to `Pending` could be problematic as mentioned in the comment above in `fail_operation_in_cloud`. An operation is `Pending` only when we have added that entry in the map. Non-existence of an entry doesn't necessarily mean `Pending`. If we receive a non-existent operation id, we can't assume that it'd in pending state and send responses to cloud.

Reply:
> I cannot come up with the cases, where the key has been removed already because of the completion but dropping into this function.

Yeah, for pretty much all the cases that I could think of, like duplicate responses from child devices, we have guard rails at other places to prevent this case from happening. As you rightly pointed out, this code can be simplified a lot, avoiding all those guard rails spread all over the place, once we have access to request IDs from the cloud. This is fine for now.